File: app/src/main/java/knf/kuma/pojos/DirectoryPage.java
package knf.kuma.pojos;

import android.content.Context;
import android.util.Log;

import org.jsoup.Jsoup;

import java.util.ArrayList;
import java.util.List;

import knf.kuma.commons.BypassUtil;
import knf.kuma.commons.Network;
import knf.kuma.database.dao.AnimeDAO;
import pl.droidsonroids.jspoon.Jspoon;
import pl.droidsonroids.jspoon.annotation.Selector;

public class DirectoryPage {
    @Selector(value = "article.Anime.alt.B > a", attr = "href")
    public List<String> links = new ArrayList<>();

    public List<AnimeObject> getAnimes(Context context, AnimeDAO animeDAO, Jspoon jspoon, UpdateInterface updateInterface) {
        List<AnimeObject> animeObjects = new ArrayList<>();
        for (String link : links) {
            if (Network.isConnected()) {
                if (!animeDAO.existLink("https://animeflv.net" + link))
                    try {
                        AnimeObject.WebInfo webInfo = jspoon.adapter(AnimeObject.WebInfo.class).fromHtml(Jsoup.connect("https://animeflv.net" + link).cookies(BypassUtil.getMapCookie(context)).userAgent(BypassUtil.userAgent).timeout(3000).get().outerHtml());
                        animeObjects.add(new AnimeObject("https://animeflv.net" + link, webInfo));
                        Log.e("Directory Getter", "Added: https://animeflv.net" + link);
                        updateInterface.onAdd();
                    } catch (Exception e) {
                        Log.e("Directory Getter", "Error adding: https://animeflv.net" + link +"\nCause: "+e.getMessage());
                        updateInterface.onError();
                    }
            } else {
                Log.e("Directory Getter", "Abort: No internet");
                break;
            }
        }
        return animeObjects;
    }

    public List<AnimeObject> getAnimesRecreate(Context context, Jspoon jspoon, UpdateInterface updateInterface) {
        List<AnimeObject> animeObjects = new ArrayList<>();
        for (String link : links) {
            if (Network.isConnected()) {
                try {
                    AnimeObject.WebInfo webInfo = jspoon.adapter(AnimeObject.WebInfo.class).fromHtml(Jsoup.connect("https://animeflv.net" + link).cookies(BypassUtil.getMapCookie(context)).userAgent(BypassUtil.userAgent).timeout(3000).get().outerHtml());
                    animeObjects.add(new AnimeObject("https://animeflv.net" + link, webInfo));
                    Log.e("Directory Getter", "Replaced: https://animeflv.net" + link);
                    updateInterface.onAdd();
                } catch (Exception e) {
                    Log.e("Directory Getter", "Error replacing: https://animeflv.net" + link +"\nCause: "+e.getMessage());
                    updateInterface.onError();
                }
            } else {
                Log.e("Directory Getter", "Abort: No internet");
                break;
            }
        }
        return animeObjects;
    }

    public interface UpdateInterface {
        void onAdd();

        void onError();
    }
}

File: app/src/main/java/knf/kuma/pojos/DirectoryPage.kt
package knf.kuma.pojos

import android.util.Log
import knf.kuma.commons.Network
import knf.kuma.commons.PrefsUtil
import knf.kuma.commons.isFullMode
import knf.kuma.commons.jsoupCookies
import knf.kuma.database.CacheDB
import knf.kuma.database.dao.AnimeDAO
import pl.droidsonroids.jspoon.Jspoon
import pl.droidsonroids.jspoon.annotation.Selector

class DirectoryPage {
    @Selector(value = "article.Anime.alt.B a.Button.Vrnmlk", attr = "href")
    var links: List<String> = listOf()

    fun getAnimes(animeDAO: AnimeDAO, jspoon: Jspoon, updateInterface: UpdateInterface, isCloudflareActive: Boolean): List<AnimeObject> {
        val animeObjects = ArrayList<AnimeObject>()
        for (link in links) {
            if (Network.isConnected) {
                if (!animeDAO.existLink("%animeflv.net$link")) {
                    try {
                        val response = jsoupCookies("https://animeflv.net$link", true).execute()
                        val body = response.body()
                        if (response.statusCode() == 200 && body != null) {
                            val webInfo =
                                jspoon.adapter(AnimeObject.WebInfo::class.java).fromHtml(body)
                            if (isFullMode && !PrefsUtil.isFamilyFriendly) {
                                animeObjects.add(AnimeObject("https://animeflv.net$link", webInfo))
                                Log.e("Directory Getter", "Added: https://animeflv.net$link")
                            } else {
                                if (webInfo.genres.contains("Ecchi"))
                                    Log.e("Directory Getter", "Skip: https://animeflv.net$link")
                                else {
                                    animeObjects.add(
                                        AnimeObject(
                                            "https://animeflv.net$link",
                                            webInfo
                                        )
                                    )
                                    Log.e("Directory Getter", "Added: https://animeflv.net$link")
                                }
                            }
                            updateInterface.onAdd()
                        } else if (response.statusCode() == 404) {
                            CacheDB.INSTANCE.animeDAO().allLinksInEmission
                        } else check(response.statusCode() < 400) { "Response code: ${response.statusCode()}" }
                    } catch (e: Exception) {
                        e.printStackTrace()
                        Log.e("Directory Getter", "Error adding: https://animeflv.net" + link + "\nCause: " + e.message)
                        updateInterface.onError()
                    }
                    if (isCloudflareActive)
                        Thread.sleep(5000)
                }
            } else {
                Log.e("Directory Getter", "Abort: No internet")
                break
            }
        }
        return animeObjects
    }

    fun getAnimesRecreate(jspoon: Jspoon, updateInterface: UpdateInterface, isCloudflareActive: Boolean): List<AnimeObject> {
        val animeObjects = ArrayList<AnimeObject>()
        for (link in links) {
            if (Network.isConnected) {
                try {
                    val webInfo = jspoon.adapter(AnimeObject.WebInfo::class.java).fromHtml(jsoupCookies("https://animeflv.net$link").get().outerHtml())
                    if (isFullMode && !PrefsUtil.isFamilyFriendly) {
                        animeObjects.add(AnimeObject("https://animeflv.net$link", webInfo))
                        Log.e("Directory Getter", "Replaced: https://animeflv.net$link")
                    } else {
                        if (webInfo.genres.contains("Ecchi"))
                            Log.e("Directory Getter", "Skip: https://animeflv.net$link")
                        else {
                            animeObjects.add(AnimeObject("https://animeflv.net$link", webInfo))
                            Log.e("Directory Getter", "Replaced: https://animeflv.net$link")
                        }
                    }
                    updateInterface.onAdd()
                } catch (e: Exception) {
                    e.printStackTrace()
                    Log.e("Directory Getter", "Error replacing: https://animeflv.net" + link + "\nCause: " + e.message)
                    updateInterface.onError()
                }
                if (isCloudflareActive)
                    Thread.sleep(5000)
            } else {
                Log.e("Directory Getter", "Abort: No internet")
                break
            }
        }
        return animeObjects
    }

    interface UpdateInterface {
        fun onAdd()

        fun onError()
    }
}
